template.yml

```diff
@@ -157,7 +224,40 @@ objects:
         - name: config
           configMap:
             name: ${NAME}-config
+        - name: ${NAME}-proxy-tls
+          secret:
+            secretName: ${NAME}-proxy-tls
+        - name: ${NAME}-proxy-cookie
+          secret:
+            secretName: ${NAME}-proxy-cookie
         containers:
+        - name: oauth-proxy
+          image: openshift/oauth-proxy:v1.1.0
+          imagePullPolicy: IfNotPresent
+          ports:
+          - containerPort: 8443
+            name: public
+          args:
+          - --https-address=:8443
+          - --provider=openshift
+          - --openshift-service-account=${NAME}
+          - --upstream=http://localhost:9099
+          - --tls-cert=/etc/tls/private/tls.crt
+          - --tls-key=/etc/tls/private/tls.key
+          - -http-address=
+          - -email-domain=*
+          - '-openshift-sar={"resource": "", "verb": "get"}'
```

Comment:
Just some random thought but have you tried this?

```
-openshift-sar={"resource": "secrets", "verb": "get", "resourceName": "autoheal-access-key", "namespace": "openshift-autoheal"}
```

I've used a similar approach in a template of mine and it worked fine (see [here](https://github.com/simonpasquier/openshift_templates/blob/f1e5f2789804406b533964e1b0a9a7e95997b1d9/prometheus_ha/prometheus.yaml#L172)).

AFAIU service accounts use tokens so they are only affected by `openshift-sar`, not `openshift-delegate-urls`.

Reply:
Tried that, no luck.

Reply:
With `openshift-delegate-urls`, I'm getting `authorizer reason: User "system:serviceaccount:openshift-monitoring:alertmanager-main" cannot get secrets in project "openshift-autoheal"`

Without it (and with the SAR mentioned by @simonpasquier ) I don't get *any* log lines in oauth-proxy, but Alertmanager still gets http 403.


`oc get clusterrole autoheal-access -o yaml`
```yaml

apiVersion: authorization.openshift.io/v1
kind: ClusterRole
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"rbac.authorization.k8s.io/v1","kind":"ClusterRole","metadata":{"annotations":{},"name":"autoheal-access","namespace":""},"rules":[{"apiGroups":[""],"resourceNames":["autoheal-access-key"],"resources":["secrets"],"verbs":["get"]}]}
  creationTimestamp: 2018-04-24T09:32:02Z
  name: autoheal-access
  resourceVersion: "3838"
  selfLink: /apis/authorization.openshift.io/v1/clusterroles/autoheal-access
  uid: 58240e0d-47a2-11e8-aede-42010a8e0002
rules:
- apiGroups:
  - ""
  attributeRestrictions: null
  resourceNames:
  - autoheal-access-key
  resources:
  - secrets
  verbs:
  - get
```

`oc get rolebinding alertmanager-autoheal-access -o yaml`
```yaml
apiVersion: authorization.openshift.io/v1
groupNames: null
kind: RoleBinding
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"rbac.authorization.k8s.io/v1","kind":"RoleBinding","metadata":{"annotations":{},"name":"alertmanager-autoheal-access","namespace":"openshift-autoheal"},"roleRef":{"apiGroup":"rbac.authorization.k8s.io","kind":"ClusterRole","name":"autoheal-access"},"subjects":[{"kind":"ServiceAccount","name":"alertmanager-main","namespace":"openshift-monitoring"}]}
  creationTimestamp: 2018-04-24T09:32:02Z
  name: alertmanager-autoheal-access
  namespace: openshift-autoheal
  resourceVersion: "3839"
  selfLink: /apis/authorization.openshift.io/v1/namespaces/openshift-autoheal/rolebindings/alertmanager-autoheal-access
  uid: 586f8e4a-47a2-11e8-aede-42010a8e0002
roleRef:
  name: autoheal-access
subjects:
- kind: ServiceAccount
  name: alertmanager-main
  namespace: openshift-monitoring
userNames:
- system:serviceaccount:openshift-monitoring:alertmanager-main
```

Could this be because one of them is a `RoleBinding` and the other is `ClusterRole`? Should I change this to `ClusterRoleBinding`? 

```
 oc policy who-can get secret autoheal-access-key -n openshift-autoheal --v=8
I0424 10:52:55.146753    3139 loader.go:357] Config loaded from file /root/.kube/config
I0424 10:52:55.166488    3139 request.go:874] Request Body: {"kind":"LocalResourceAccessReview","apiVersion":"authorization.openshift.io/v1","namespace":"","verb":"get","resourceAPIGroup":"","resourceAPIVersion":"","resource":"secrets","resourceName":"autoheal-access-key","path":"","isNonResourceURL":false,"content":null}
I0424 10:52:55.166592    3139 round_trippers.go:383] POST https://internal-api.elad-dev.origin-gce.dev.openshift.com:8443/apis/authorization.openshift.io/v1/namespaces/openshift-autoheal/localresourceaccessreviews
I0424 10:52:55.166622    3139 round_trippers.go:390] Request Headers:
I0424 10:52:55.166632    3139 round_trippers.go:393]     Accept: application/json, */*
I0424 10:52:55.166641    3139 round_trippers.go:393]     Content-Type: application/json
I0424 10:52:55.166649    3139 round_trippers.go:393]     User-Agent: oc/v1.10.0+b81c8f8 (linux/amd64) kubernetes/b81c8f8
I0424 10:52:55.180299    3139 round_trippers.go:408] Response Status: 201 Created in 13 milliseconds
I0424 10:52:55.180336    3139 round_trippers.go:411] Response Headers:
I0424 10:52:55.180346    3139 round_trippers.go:414]     Content-Type: application/json
I0424 10:52:55.180354    3139 round_trippers.go:414]     Content-Length: 1189
I0424 10:52:55.180362    3139 round_trippers.go:414]     Date: Tue, 24 Apr 2018 10:52:55 GMT
I0424 10:52:55.180372    3139 round_trippers.go:414]     Cache-Control: no-store
I0424 10:52:55.180414    3139 request.go:874] Response Body: {"kind":"ResourceAccessReviewResponse","apiVersion":"authorization.openshift.io/v1","namespace":"openshift-autoheal","users":["system:admin","system:kube-controller-manager","system:serviceaccount:kube-system:clusterrole-aggregation-controller","system:serviceaccount:kube-system:generic-garbage-collector","system:serviceaccount:kube-system:namespace-controller","system:serviceaccount:kube-system:persistent-volume-binder","system:serviceaccount:openshift-infra:build-controller","system:serviceaccount:openshift-infra:cluster-quota-reconciliation-controller","system:serviceaccount:openshift-infra:ingress-to-route-controller","system:serviceaccount:openshift-infra:service-serving-cert-controller","system:serviceaccount:openshift-infra:serviceaccount-pull-secrets-controller","system:serviceaccount:openshift-infra:template-instance-controller","system:serviceaccount:openshift-infra:template-service-broker","system:serviceaccount:openshift-monitoring:alertmanager-main","system:serviceaccount:openshift-monitoring:clu [truncated 165 chars]
Namespace: openshift-autoheal
Verb:      get
Resource:  secrets

Users:  system:admin
        system:kube-controller-manager
        system:serviceaccount:kube-system:clusterrole-aggregation-controller
        system:serviceaccount:kube-system:generic-garbage-collector
        system:serviceaccount:kube-system:namespace-controller
        system:serviceaccount:kube-system:persistent-volume-binder
        system:serviceaccount:openshift-infra:build-controller
        system:serviceaccount:openshift-infra:cluster-quota-reconciliation-controller
        system:serviceaccount:openshift-infra:ingress-to-route-controller
        system:serviceaccount:openshift-infra:service-serving-cert-controller
        system:serviceaccount:openshift-infra:serviceaccount-pull-secrets-controller
        system:serviceaccount:openshift-infra:template-instance-controller
        system:serviceaccount:openshift-infra:template-service-broker
        system:serviceaccount:openshift-monitoring:alertmanager-main
        system:serviceaccount:openshift-monitoring:cluster-monitoring-operator
        system:serviceaccount:openshift-monitoring:prometheus-operator

Groups: system:cluster-admins
        system:masters
```
And when logged in as the alertmanager-main serviceaccount:
``` 
# oc auth can-i get secret/autoheal-access-key -n openshift-autoheal
yes
```

Reply:
Yes, I think that the problem is that OpenShift is ignoring the kind of role that you specified in the role binding. You said this:

```yaml
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: autoheal-access
```

But it only remembered this:

```yaml
roleRef:
  name: autoheal-access
```

So that role binding is only applying to a non-existing `autoheal-access` role inside the `openshift-autoheal` namespace.

When I tested this I did it in a Kubernetes 3.10 cluster. It may be that the behaviour has not been backported, or that it just isn't available in OpenShift 3.10.

@ironcladlou do you know who can best help us with this issue?

@elad661 if the above is accurate, and if you are loggged in with the alert manager service account then the response should have been `no`. Are you sure you used that service account? Can you share the `--v=8` output of that command?

Reply:
```
# oc whoami
system:serviceaccount:openshift-monitoring:alertmanager-main
```

```
# oc auth can-i get secret/autoheal-access-key -n openshift-autoheal --v=8
I0424 11:09:50.783564   10534 loader.go:357] Config loaded from file /root/.kube/config
I0424 11:09:50.803465   10534 request.go:874] Request Body: {"kind":"SelfSubjectAccessReview","apiVersion":"authorization.k8s.io/v1","metadata":{"creationTimestamp":null},"spec":{"resourceAttributes":{"namespace":"openshift-autoheal","verb":"get","resource":"secrets","name":"autoheal-access-key"}},"status":{"allowed":false}}
I0424 11:09:50.803568   10534 round_trippers.go:383] POST https://internal-api.elad-dev.origin-gce.dev.openshift.com:8443/apis/authorization.k8s.io/v1/selfsubjectaccessreviews
I0424 11:09:50.803581   10534 round_trippers.go:390] Request Headers:
I0424 11:09:50.803590   10534 round_trippers.go:393]     Accept: application/json, */*
I0424 11:09:50.803600   10534 round_trippers.go:393]     Content-Type: application/json
I0424 11:09:50.803616   10534 round_trippers.go:393]     User-Agent: oc/v1.10.0+b81c8f8 (linux/amd64) kubernetes/b81c8f8
I0424 11:09:50.803628   10534 round_trippers.go:393]     Authorization: Bearer eyJhbGciOiJSUzI1NiIsImtpZCI6IiJ9.eyJpc3MiOiJrdWJlcm5ldGVzL3NlcnZpY2VhY2NvdW50Iiwia3ViZXJuZXRlcy5pby9zZXJ2aWNlYWNjb3VudC9uYW1lc3BhY2UiOiJvcGVuc2hpZnQtbW9uaXRvcmluZyIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VjcmV0Lm5hbWUiOiJhbGVydG1hbmFnZXItbWFpbi10b2tlbi04NnhtNSIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50Lm5hbWUiOiJhbGVydG1hbmFnZXItbWFpbiIsImt1YmVybmV0ZXMuaW8vc2VydmljZWFjY291bnQvc2VydmljZS1hY2NvdW50LnVpZCI6ImI2MTQ0NDc3LTQ3YTEtMTFlOC1hZWRlLTQyMDEwYThlMDAwMiIsInN1YiI6InN5c3RlbTpzZXJ2aWNlYWNjb3VudDpvcGVuc2hpZnQtbW9uaXRvcmluZzphbGVydG1hbmFnZXItbWFpbiJ9.i7MhpIbZ2T2RZVYXBIZLtcJSuRZ6z5lNOiSbwe_w7vaB3lyapstX91EtnjlyY58whV4vJyg5E5yynhSBX7XLof3qvNQnkHbtLdXFw8GK2GnlY1jm6sSKq5wzZOm99SjYHGgqmA_IQntrwfCJHiUpUmnFMLF8BNKEMj_Rvsub620Ed6kIEQJh09HMGH8ciVCKROBQSTWRCX19tuAHfqA32HxW5AxFiplmKK8KEvnFbQqPCg64pD8B_tbqNBBLTPUprPXly1kP04GJJbIwq6K57K-qf8zf7l3wUZee7aIzvCv2Xjr-dyT_VxVXjnUcHwtGqzso_hS9_9JBxj45W2ELrA
I0424 11:09:50.817093   10534 round_trippers.go:408] Response Status: 201 Created in 13 milliseconds
I0424 11:09:50.817157   10534 round_trippers.go:411] Response Headers:
I0424 11:09:50.817167   10534 round_trippers.go:414]     Date: Tue, 24 Apr 2018 11:09:50 GMT
I0424 11:09:50.817176   10534 round_trippers.go:414]     Cache-Control: no-store
I0424 11:09:50.817188   10534 round_trippers.go:414]     Content-Type: application/json
I0424 11:09:50.817196   10534 round_trippers.go:414]     Content-Length: 309
I0424 11:09:50.817242   10534 request.go:874] Response Body: {"kind":"SelfSubjectAccessReview","apiVersion":"authorization.k8s.io/v1","metadata":{"creationTimestamp":null},"spec":{"resourceAttributes":{"namespace":"openshift-autoheal","verb":"get","resource":"secrets","name":"autoheal-access-key"}},"status":{"allowed":true,"reason":"allowed by openshift authorizer"}}
yes
```

Reply:
aha! it should be `name`, not `resourceName`!